core/server/lib/members/static/gateway/bundle.js

```diff
@@ -23,9 +23,76 @@
         };
     }
 
+    function isTokenExpired(token) {
+        try {
+            const [header, claims, signature] = token.split('.');
+
+            const parsedClaims = JSON.parse(atob(claims.replace('+', '-').replace('/', '_')));
+
+            const expiry = parsedClaims.exp * 1000;
+            const now = Date.now();
+
+            const nearFuture = now + (30 * 1000);
+
+            if (expiry > nearFuture) {
+                return true;
+            }
+
+            return false;
+        } catch (e) {
+            return true;
+        }
+    }
+
+    function getStoredToken(audience) {
+        const tokenKey = 'members:token:aud:' + audience;
+        const storedToken = storage.getItem(tokenKey);
+        if (isTokenExpired(storedToken)) {
+            storage.removeItem(tokenKey);
+            return null;
+        }
+        return storedToken;
+    }
+
+    function getStoredTokenKeys() {
+        try {
+            return JSON.parse(storage.getItem('members:tokens') || '[]');
+        } catch (e) {
+            storage.removeItem('members:tokens');
+            return [];
+        }
+    }
+
+    function addStoredToken(audience, token) {
+        const storedTokenKeys = getStoredTokenKeys();
+        const tokenKey = 'members:token:aud:' + audience;
+
+        storage.setItem(tokenKey, token);
+        if (!storedTokenKeys.includes(tokenKey)) {
+            storage.setItem('members:tokens', JSON.stringify(storedTokenKeys.concat(tokenKey)));
+        }
+    }
+
+    function clearStorage() {
+        storage.removeItem('signedin');
+        const storedTokenKeys = getStoredTokenKeys();
+
+        storedTokenKeys.forEach(function (key) {
```

Comment:
This bs is because we share localstorage with the rest of the frontend. If any themes/injected scripts use localstorage, we could break things if we `storage.clear()`